beacon-chain/rpc/lookup/blocker.go

```diff
@@ -132,6 +136,126 @@ func (p *BeaconDbBlocker) Block(ctx context.Context, id []byte) (interfaces.Read
 	return blk, nil
 }
 
+// blobsFromBlobs retrieves blobs corresponding to `indices` and `root` from the store, expecting blobs to be
+// stored directly (aka. no data columns).
+func (p *BeaconDbBlocker) blobsFromBlobs(indices map[uint64]bool, root []byte) ([]*blocks.VerifiedROBlob, *core.RpcError) {
+	// If no indices are provided in the request, we fetch all available blobs for the block.
+	if len(indices) == 0 {
+		// Get all blob indices for the block.
+		indicesMap, err := p.BlobStorage.Indices(bytesutil.ToBytes32(root))
+		if err != nil {
+			log.WithField("blockRoot", hexutil.Encode(root)).Error(errors.Wrapf(err, "could not retrieve blob indices for root %#x", root))
+			return nil, &core.RpcError{Err: fmt.Errorf("could not retrieve blob indices for root %#x", root), Reason: core.Internal}
+		}
+
+		for indice, exists := range indicesMap {
+			if exists {
+				indices[uint64(indice)] = true
+			}
+		}
+	}
+
+	// Retrieve from the store the blobs corresponding to the indices for this block root.
+	blobs := make([]*blocks.VerifiedROBlob, 0, len(indices))
+	for index := range indices {
+		vblob, err := p.BlobStorage.Get(bytesutil.ToBytes32(root), index)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"blockRoot": hexutil.Encode(root),
+				"blobIndex": index,
+			}).Error(errors.Wrapf(err, "could not retrieve blob for block root %#x at index %d", root, index))
+			return nil, &core.RpcError{Err: fmt.Errorf("could not retrieve blob for block root %#x at index %d", root, index), Reason: core.Internal}
+		}
+		blobs = append(blobs, &vblob)
+	}
+
+	return blobs, nil
+}
+
+// blobsFromDataColumns retrieves data columns from the store, convert them to blobs, and return blobs corresponding to `indices` and `root` from the store,
+// expecting data columns to be stored (aka. no blobs).
+// If not all data columns are available, the function returns a "not found" error.
+// This function expects the block associated with root has blobs.
+func (p *BeaconDbBlocker) blobsFromDataColumns(indices map[uint64]bool, rootBytes []byte) ([]*blocks.VerifiedROBlob, *core.RpcError) {
+	// Multiple implementations are possible here, depending on the data storage strategy.
```

Comment:
disagree with this reasoning, we shouldn't only support 1. If a node has the required amount of columns we should return the full blob rather than error out if they aren't subscribed to all the subnets. They can also recover all the blobs if they are subscribed to half the data column subnets, so having this as a requirement doesn't make sense.

If we are worried about inconsistent use of the API, we can return an error where we mention that a user can run a node with `--subscribe-all-subnets` flag if they need consistent API results.

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/pull/14312/commits/56b275bf0cd6006efda6d57a1da688ba639cf340.